sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsShuffleHeartbeatManager.scala

```diff
@@ -79,30 +115,67 @@ class RapidsShuffleHeartbeatManager extends Logging {
    * @return `RapidsExecutorUpdateMsg` with new executors, since the last heartbeat was received.
    */
   def executorHeartbeat(id: BlockManagerId): RapidsExecutorUpdateMsg = synchronized {
-    val lastRegistration = lastRegistrationSeen.get(id)
+    removeDeadExecutors(getCurrentTimeMillis)
+    val lastRegistration = executorRegistrations.get(id)
     if (lastRegistration == null) {
-      throw new IllegalStateException(s"Heartbeat from unknown executor $id")
+      logDebug(s"Heartbeat from unknown executor $id")
+      registerExecutor(id)
```

Comment:
Are executors robust in terms of being notified of executors they already know about (including themselves?!). Glancing at the code, it looks like they will blindly attempt to connect to each executor reported in the message.

Reply:
On `registerExecutor` we get the collection of executors to send to the registrant _before_ adding ourselves. So it doesn't happen, since we ended up in `registerExecutor` because we are not in the list. That said, I should make this more robust. The transport should refuse (and blow up) if it is trying to connect to itself, there is no reason to do that. Let me know if you'd like to see that in this PR.

The heartbeat endpoint, on update, ends up calling `https://github.com/NVIDIA/spark-rapids/blob/branch-21.08/shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCX.scala#L787`. This protects against existing connections. This is the same function we call when we are initializing the `RapidsShuffleIterator` (i.e. after the early wakeup happens).